File: context_memory.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
import re
import yaml


def _load_model_router() -> str:
    try:
        with open("personalization.yaml", "r", encoding="utf-8") as f:
            raw = yaml.safe_load(f) or {}
            return (raw.get("models", {}) or {}).get("router", "qwen-max")
    except Exception:
        return "qwen-max"


ROUTER_MODEL = _load_model_router()


@dataclass
class MemoryStore:
    """长期记忆与每日日志存储。"""

    root_dir: Path

    def __post_init__(self):
        self.memory_md = self.root_dir / "MEMORY.md"
        self.daily_dir = self.root_dir / "memory"

    def ensure_md_files(self) -> None:
        self.root_dir.mkdir(parents=True, exist_ok=True)
        self.daily_dir.mkdir(parents=True, exist_ok=True)
        if not self.memory_md.exists():
            self.memory_md.write_text(
                "# MEMORY\n\n"
                "## 长期稳定信息（偏好/规则/身份/项目约定）\n"
                "- 你是一个可靠、务实、尊重用户意图的 AI 助手。\n\n"
                "## 稳定偏好提炼\n"
                "- 暂无。\n\n"
                "## 归档摘要\n"
                "- 暂无。\n",
                encoding="utf-8",
            )
        today_path = self.get_daily_file_path()
        if not today_path.exists():
            today_path.write_text(
                f"# {datetime.now().strftime('%Y-%m-%d')} 工作记忆\n\n"
                "## 今天做了什么\n\n"
                "## 临时决定\n\n"
                "## 正在排查的问题\n\n"
                "## 对话记录\n",
                encoding="utf-8",
            )

    def read_memory(self) -> str:
        return self._read_text(self.memory_md)

    def write_memory(self, memory_text: str) -> None:
        self.memory_md.write_text(memory_text, encoding="utf-8")

    def get_daily_file_path(self, day: datetime | None = None) -> Path:
        day = day or datetime.now()
        return self.daily_dir / f"{day.strftime('%Y-%m-%d')}.md"

    def append_daily_dialogue(self, user_input: str, assistant_output: str) -> None:
        path = self.get_daily_file_path()
        if not path.exists():
            self.ensure_md_files()
        now = datetime.now().strftime("%Y-%m-%d %H:%M")
        safe_user = user_input.replace("\n", " ").strip()
        safe_assistant = assistant_output.replace("\n", " ").strip()
        line = f"- [{now}] U: {safe_user} | A: {safe_assistant[:300]}\n"
        with path.open("a", encoding="utf-8") as f:
            f.write(line)

    def read_recent_daily_lines(self, limit: int = 40) -> list[str]:
        if not self.daily_dir.exists():
            return []
        collected: list[str] = []
        for p in sorted(self.daily_dir.glob("*.md"), reverse=True):
            text = p.read_text(encoding="utf-8")
            lines = [ln for ln in text.splitlines() if ln.strip().startswith("- [")]
            collected.extend(lines[::-1])
            if len(collected) >= limit:
                break
        return collected[:limit]

    def append_memory_section(self, section_title: str, content: str) -> None:
        memory_text = self.read_memory()
        marker = f"## {section_title}"
        if marker not in memory_text:
            memory_text = f"{memory_text.rstrip()}\n\n{marker}\n"
        memory_text = f"{memory_text.rstrip()}\n- {content.strip()}\n"
        self.write_memory(memory_text)

    @staticmethod
    def _read_text(path: Path) -> str:
        return path.read_text(encoding="utf-8").strip()


@dataclass
class MemoryRefiner:
    """负责提取/整理对话为 memory，并把稳定偏好提炼到 soul。"""

    client: object
    store: MemoryStore
    max_recent_lines: int = 40

    def update_memory(self, user_input: str, assistant_output: str) -> None:
        self.store.append_daily_dialogue(user_input, assistant_output)

        recent_lines = self.store.read_recent_daily_lines(limit=self.max_recent_lines + 8)
        if len(recent_lines) > self.max_recent_lines:
            cut = len(recent_lines) // 2
            compressed = self._compress_with_llm(recent_lines[:cut])
            day = datetime.now().strftime("%Y-%m-%d")
            self.store.append_memory_section("归档摘要", f"{day} 压缩记忆: {compressed}")

    def maybe_update_soul(self) -> None:
        recent_lines = self.store.read_recent_daily_lines(limit=24)
        if len(recent_lines) == 0 or len(recent_lines) % 8 != 0:
            return

        memory_text = self.store.read_memory()
        recent_context = "\n".join(recent_lines[-12:])
        try:
            response = self.client.chat.completions.create(
                model=ROUTER_MODEL,
                messages=[
                    {
                        "role": "system",
                        "content": "你是长期偏好提炼器。根据最近对话提炼稳定偏好/规则/项目约定，输出 3-6 条短句。",
                    },
                    {
                        "role": "user",
                        "content": f"当前 MEMORY:\n{memory_text}\n\n最近对话:\n{recent_context}",
                    },
                ],
                temperature=0.2,
                stream=False,
            )
            updated = (response.choices[0].message.content or "").strip()
            if updated:
                self.store.append_memory_section("稳定偏好提炼", updated)
        except Exception:
            pass

    def compact_history_if_needed(self, history_list: list, max_chars: int = 12000) -> list:
        """
        当上下文接近上限时自动压缩历史，进入下一段对话。
        返回压缩后的新 history_list（保留近期消息，老消息写入 memory.md）。
        """
        total_chars = sum(len((m.get("content") or "")) for m in history_list if isinstance(m, dict))
        if total_chars < max_chars or len(history_list) < 12:
            return history_list

        cut = int(len(history_list) * 0.7)
        old_chunk = history_list[:cut]
        keep_chunk = history_list[cut:]

        compact_source = []
        for msg in old_chunk:
            role = msg.get("role", "")
            content = (msg.get("content") or "").replace("\n", " ")
            compact_source.append(f"{role}: {content[:300]}")

        summary = self._compress_with_llm(compact_source)
        self.update_memory("系统自动压缩上下文", summary)
        return keep_chunk

    def _compress_with_llm(self, lines: list[str]) -> str:
        joined = "\n".join(lines)
        try:
            response = self.client.chat.completions.create(
                model=ROUTER_MODEL,
                messages=[
                    {
                        "role": "system",
                        "content": "你是记忆压缩器。请把对话压缩成 3-6 条可复用记忆，聚焦偏好、目标、约束。",
                    },
                    {"role": "user", "content": joined},
                ],
                temperature=0.2,
                stream=False,
            )
            return (response.choices[0].message.content or "").strip()
        except Exception:
            preview = " | ".join(lines[:3])
            return f"- 历史对话压缩（fallback）: {preview[:200]}"


@dataclass
class ContextAssembler:
    """负责拼接模型需要的系统上下文。"""

    store: MemoryStore

    def build_system_context(self) -> str:
        memory = self.store.read_memory()
        daily_lines = self.store.read_recent_daily_lines(limit=20)
        recent_daily = "\n".join(daily_lines[-12:]) if daily_lines else "- 暂无今日记录"
        recent_preview = memory[-2400:] if len(memory) > 2400 else memory
        return (
            "请严格遵循以下长期上下文（MEMORY.md）与最近工作日志（memory/YYYY-MM-DD.md）：\n\n"
            f"{recent_preview}\n\n## 最近工作日志片段\n{recent_daily}"
        )


@dataclass
class MemoryContextManager:
    """
    兼容旧接口的门面类，内部拆分为：
    - MemoryStore（长期记忆存储）
    - MemoryRefiner（记忆提取/整理/提炼）
    - ContextAssembler（模型上下文拼接）

    支持 user_scope_id：每个用户范围独立长期记忆目录。
    """

    client: object
    root_dir: Path
    max_recent_lines: int = 40

    def __post_init__(self):
        self.memory_scopes_dir = self.root_dir / "memory_scopes"
        self._stores: dict[str, MemoryStore] = {}
        self._refiners: dict[str, MemoryRefiner] = {}
        self._assemblers: dict[str, ContextAssembler] = {}

    @staticmethod
    def _sanitize_scope_id(user_scope_id: str | None) -> str:
        raw = (user_scope_id or "default").strip() or "default"
        return re.sub(r"[^a-zA-Z0-9._-]", "_", raw)

    def _ensure_scope_components(self, user_scope_id: str | None):
        scope = self._sanitize_scope_id(user_scope_id)
        if scope not in self._stores:
            scope_dir = self.memory_scopes_dir / scope
            store = MemoryStore(root_dir=scope_dir)
            refiner = MemoryRefiner(
                client=self.client,
                store=store,
                max_recent_lines=self.max_recent_lines,
            )
            assembler = ContextAssembler(store=store)
            self._stores[scope] = store
            self._refiners[scope] = refiner
            self._assemblers[scope] = assembler
        return scope

    def ensure_md_files(self, user_scope_id: str | None = None) -> None:
        scope = self._ensure_scope_components(user_scope_id)
        self._stores[scope].ensure_md_files()

    def build_system_context(self, user_scope_id: str | None = None) -> str:
        scope = self._ensure_scope_components(user_scope_id)
        self._stores[scope].ensure_md_files()
        return self._assemblers[scope].build_system_context()

    def update_memory(self, user_input: str, assistant_output: str, user_scope_id: str | None = None) -> None:
        scope = self._ensure_scope_components(user_scope_id)
        self._stores[scope].ensure_md_files()
        self._refiners[scope].update_memory(user_input, assistant_output)

    def maybe_update_soul(self, user_scope_id: str | None = None) -> None:
        scope = self._ensure_scope_components(user_scope_id)
        self._stores[scope].ensure_md_files()
        self._refiners[scope].maybe_update_soul()

    def compact_history_if_needed(
        self,
        history_list: list,
        max_chars: int = 12000,
        user_scope_id: str | None = None,
    ) -> list:
        scope = self._ensure_scope_components(user_scope_id)
        self._stores[scope].ensure_md_files()
        return self._refiners[scope].compact_history_if_needed(history_list, max_chars=max_chars)

File: src/context-memory.ts
import fs from "node:fs";
import path from "node:path";
import type { ChatHistoryItem } from "./types.js";
import { loadModels } from "./config.js";

function nowDay(): string {
  return new Date().toISOString().slice(0, 10);
}

function nowMinuteLocal(): string {
  const d = new Date();
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ${pad(d.getHours())}:${pad(d.getMinutes())}`;
}

function readText(pathname: string): string {
  return fs.existsSync(pathname) ? fs.readFileSync(pathname, "utf8").trim() : "";
}

export class MemoryStore {
  readonly memoryMd: string;
  readonly dailyDir: string;

  constructor(readonly rootDir: string) {
    this.memoryMd = path.join(rootDir, "MEMORY.md");
    this.dailyDir = path.join(rootDir, "memory");
  }

  ensureMdFiles(): void {
    fs.mkdirSync(this.rootDir, { recursive: true });
    fs.mkdirSync(this.dailyDir, { recursive: true });

    if (!fs.existsSync(this.memoryMd)) {
      fs.writeFileSync(
        this.memoryMd,
        [
          "# MEMORY",
          "",
          "## 长期稳定信息（偏好/规则/身份/项目约定）",
          "- 你是一个可靠、务实、尊重用户意图的 AI 助手。",
          "",
          "## 稳定偏好提炼",
          "- 暂无。",
          "",
          "## 归档摘要",
          "- 暂无。",
          "",
        ].join("\n"),
        "utf8",
      );
    }

    const todayPath = this.getDailyFilePath();
    if (!fs.existsSync(todayPath)) {
      fs.writeFileSync(
        todayPath,
        [`# ${nowDay()} 工作记忆`, "", "## 今天做了什么", "", "## 临时决定", "", "## 正在排查的问题", "", "## 对话记录", ""].join("\n"),
        "utf8",
      );
    }
  }

  readMemory(): string {
    return readText(this.memoryMd);
  }

  writeMemory(memoryText: string): void {
    fs.writeFileSync(this.memoryMd, memoryText, "utf8");
  }

  getDailyFilePath(day = new Date()): string {
    const y = day.getFullYear();
    const m = String(day.getMonth() + 1).padStart(2, "0");
    const d = String(day.getDate()).padStart(2, "0");
    return path.join(this.dailyDir, `${y}-${m}-${d}.md`);
  }

  appendDailyDialogue(userInput: string, assistantOutput: string): void {
    const dailyPath = this.getDailyFilePath();
    if (!fs.existsSync(dailyPath)) {
      this.ensureMdFiles();
    }
    const safeUser = userInput.replaceAll("\n", " ").trim();
    const safeAssistant = assistantOutput.replaceAll("\n", " ").trim();
    const line = `- [${nowMinuteLocal()}] U: ${safeUser} | A: ${safeAssistant.slice(0, 300)}\n`;
    fs.appendFileSync(dailyPath, line, "utf8");
  }

  readRecentDailyLines(limit = 40): string[] {
    if (!fs.existsSync(this.dailyDir)) {
      return [];
    }
    const collected: string[] = [];
    const files = fs
      .readdirSync(this.dailyDir)
      .filter((name) => name.endsWith(".md"))
      .sort()
      .reverse();

    for (const file of files) {
      const text = fs.readFileSync(path.join(this.dailyDir, file), "utf8");
      const lines = text.split(/\r?\n/).filter((line) => line.trim().startsWith("- ["));
      collected.push(...lines.reverse());
      if (collected.length >= limit) {
        break;
      }
    }
    return collected.slice(0, limit);
  }

  appendMemorySection(sectionTitle: string, content: string): void {
    let memoryText = this.readMemory();
    const marker = `## ${sectionTitle}`;
    if (!memoryText.includes(marker)) {
      memoryText = `${memoryText.trim()}\n\n${marker}\n`;
    }
    memoryText = `${memoryText.trim()}\n- ${content.trim()}\n`;
    this.writeMemory(memoryText);
  }
}

export class MemoryRefiner {
  private readonly routerModel: string;

  constructor(
    private readonly client: any,
    private readonly store: MemoryStore,
    private readonly maxRecentLines = 40,
    projectRoot = process.cwd(),
  ) {
    this.routerModel = loadModels(projectRoot).router;
  }

  async updateMemory(userInput: string, assistantOutput: string): Promise<void> {
    this.store.appendDailyDialogue(userInput, assistantOutput);

    const recentLines = this.store.readRecentDailyLines(this.maxRecentLines + 8);
    if (recentLines.length > this.maxRecentLines) {
      const cut = Math.floor(recentLines.length / 2);
      const compressed = await this.compressWithLlm(recentLines.slice(0, cut));
      this.store.appendMemorySection("归档摘要", `${nowDay()} 压缩记忆: ${compressed}`);
    }
  }

  async maybeUpdateSoul(): Promise<void> {
    const recentLines = this.store.readRecentDailyLines(24);
    if (recentLines.length === 0 || recentLines.length % 8 !== 0) {
      return;
    }

    const memoryText = this.store.readMemory();
    const recentContext = recentLines.slice(-12).join("\n");
    try {
      const response = await this.client.chat.completions.create({
        model: this.routerModel,
        messages: [
          {
            role: "system",
            content: "你是长期偏好提炼器。根据最近对话提炼稳定偏好/规则/项目约定，输出 3-6 条短句。",
          },
          {
            role: "user",
            content: `当前 MEMORY:\n${memoryText}\n\n最近对话:\n${recentContext}`,
          },
        ],
        temperature: 0.2,
        stream: false,
      });
      const updated = String(response.choices?.[0]?.message?.content ?? "").trim();
      if (updated) {
        this.store.appendMemorySection("稳定偏好提炼", updated);
      }
    } catch {
      // Memory refinement is best-effort and must not fail the user request.
    }
  }

  async compactHistoryIfNeeded(historyList: ChatHistoryItem[], maxChars = 12_000): Promise<ChatHistoryItem[]> {
    const totalChars = historyList.reduce((sum, msg) => {
      const content = typeof msg.content === "string" ? msg.content : "";
      return sum + content.length;
    }, 0);
    if (totalChars < maxChars || historyList.length < 12) {
      return historyList;
    }

    const cut = Math.floor(historyList.length * 0.7);
    const oldChunk = historyList.slice(0, cut);
    const keepChunk = historyList.slice(cut);

    const compactSource = oldChunk.map((msg) => {
      const role = typeof msg.role === "string" ? msg.role : "";
      const content = typeof msg.content === "string" ? msg.content.replaceAll("\n", " ") : "";
      return `${role}: ${content.slice(0, 300)}`;
    });

    const summary = await this.compressWithLlm(compactSource);
    await this.updateMemory("系统自动压缩上下文", summary);
    return keepChunk;
  }

  private async compressWithLlm(lines: string[]): Promise<string> {
    const joined = lines.join("\n");
    try {
      const response = await this.client.chat.completions.create({
        model: this.routerModel,
        messages: [
          {
            role: "system",
            content: "你是记忆压缩器。请把对话压缩成 3-6 条可复用记忆，聚焦偏好、目标、约束。",
          },
          { role: "user", content: joined },
        ],
        temperature: 0.2,
        stream: false,
      });
      return String(response.choices?.[0]?.message?.content ?? "").trim();
    } catch {
      const preview = lines.slice(0, 3).join(" | ");
      return `- 历史对话压缩（fallback）: ${preview.slice(0, 200)}`;
    }
  }
}

export class ContextAssembler {
  constructor(private readonly store: MemoryStore) {}

  buildSystemContext(): string {
    const memory = this.store.readMemory();
    const dailyLines = this.store.readRecentDailyLines(20);
    const recentDaily = dailyLines.length > 0 ? dailyLines.slice(-12).join("\n") : "- 暂无今日记录";
    const recentPreview = memory.length > 2400 ? memory.slice(-2400) : memory;
    return `请严格遵循以下长期上下文（MEMORY.md）与最近工作日志（memory/YYYY-MM-DD.md）：\n\n${recentPreview}\n\n## 最近工作日志片段\n${recentDaily}`;
  }
}

export class MemoryContextManager {
  private readonly memoryScopesDir: string;
  private readonly stores = new Map<string, MemoryStore>();
  private readonly refiners = new Map<string, MemoryRefiner>();
  private readonly assemblers = new Map<string, ContextAssembler>();

  constructor(
    private readonly client: any,
    private readonly rootDir: string,
    private readonly maxRecentLines = 40,
  ) {
    this.memoryScopesDir = path.join(rootDir, "memory_scopes");
  }

  private sanitizeScopeId(userScopeId?: string): string {
    const raw = (userScopeId ?? "default").trim() || "default";
    return raw.replace(/[^a-zA-Z0-9._-]/g, "_");
  }

  private ensureScopeComponents(userScopeId?: string): string {
    const scope = this.sanitizeScopeId(userScopeId);
    if (!this.stores.has(scope)) {
      const scopeDir = path.join(this.memoryScopesDir, scope);
      const store = new MemoryStore(scopeDir);
      const refiner = new MemoryRefiner(this.client, store, this.maxRecentLines, this.rootDir);
      const assembler = new ContextAssembler(store);
      this.stores.set(scope, store);
      this.refiners.set(scope, refiner);
      this.assemblers.set(scope, assembler);
    }
    return scope;
  }

  ensureMdFiles(userScopeId?: string): void {
    const scope = this.ensureScopeComponents(userScopeId);
    this.stores.get(scope)?.ensureMdFiles();
  }

  buildSystemContext(userScopeId?: string): string {
    const scope = this.ensureScopeComponents(userScopeId);
    this.stores.get(scope)?.ensureMdFiles();
    return this.assemblers.get(scope)?.buildSystemContext() ?? "";
  }

  async updateMemory(userInput: string, assistantOutput: string, userScopeId?: string): Promise<void> {
    const scope = this.ensureScopeComponents(userScopeId);
    this.stores.get(scope)?.ensureMdFiles();
    await this.refiners.get(scope)?.updateMemory(userInput, assistantOutput);
  }

  async maybeUpdateSoul(userScopeId?: string): Promise<void> {
    const scope = this.ensureScopeComponents(userScopeId);
    this.stores.get(scope)?.ensureMdFiles();
    await this.refiners.get(scope)?.maybeUpdateSoul();
  }

  async compactHistoryIfNeeded(
    historyList: ChatHistoryItem[],
    maxChars = 12_000,
    userScopeId?: string,
  ): Promise<ChatHistoryItem[]> {
    const scope = this.ensureScopeComponents(userScopeId);
    this.stores.get(scope)?.ensureMdFiles();
    return (await this.refiners.get(scope)?.compactHistoryIfNeeded(historyList, maxChars)) ?? historyList;
  }
}
